d3-selection/index.d.ts

```diff
@@ -845,7 +878,7 @@ interface Selection<GElement extends BaseType, Datum, PElement extends BaseType,
     /**
      * Return the first (non-null) element in this selection. If the selection is empty, returns null.
      */
-    node(): GElement | null;
+    node(): GElement;
```

Comment:
I did see your comment about this change, and I'd like to offer a better-worded critique than the one above.

Under the following assumptions:

1. TypeScript's philosophy is to catch, at compile type, potential gotchas;
2. When `strictNullChecks` is in effect, the user has requested stronger type checking than tsc currently offers; and
3. Interfacing with the DOM is almost impossible to guarantee statically;

then I think that preference should be given to _omitting_ `null` from the default selection type, and methods like `.node()` should _unconditionally_ have `| null` in their return type. This reflects a small change in the intuition, and I would argue it's the same intuition as the return type of `Local<T>.get`. `Local<T>.get()` returns `T | undefined`, which can be understood as saying "If a returned value is found, it is of type `T`". The failure mode of this lookup is captured by the `| undefined` arm.

In the case of `.node()`, the return type can be read as "If the element is found, it is of `BaseType`". The failure mode is the `| null` arm. In fact, I would go so far as to say that's the [exact interpretation of the code powering `.node()`](https://github.com/d3/d3-selection/blob/master/src/selection/node.js).

You wrote:

> IMHO, for those that go with strict null checking, it is required to have that degree of reflection, when applying the generics on the various flavors of select/selectAll, as should be the case with any casting.

I agree that you have to stop and think. But `strictNullChecks` is intended to prevent accidentally forgetting about failure modes (map/object lookups, array indexing, etc.). The problem with this change is that, when I specify a type parameter, I may have deleted a failure mode. I would argue this violates developers' expectations. Giving a type parameter to `Map<K, V>` doesn't mean that `.get()` suddenly never returns `undefined`, for instance. We want to narrow the type of `.node()` (and, by extension, the `this` context passed into accessors and `.each()`, etc.), not change semantics.

> So, for what it's worth the discipline is needed anyway, if e.g. one uses a pure class-based css-selector, which could return different element types.

This is true, but I'd argue that it's not the same thing. We're talking about two separate classes of bug:

1. The CSS selector didn't match the _right kind of element_: this is a mismatch between the type provided in the generic overload `d3.select<T>(...)`.
2. The CSS selector didn't match _at all_: this is a mismatch between the developer's expectation of the DOM and the reality.

I do realize this is a lot of words for what is ultimately a fairly small change, but it one of those things that can turn out to be fairly dangerous (at least in my mind), and it seems like a step backwards compared to all this other work making failure modes explicit.